### PR TITLE
Add Tensor.as_param()

### PR DESCRIPTION
**Pull request by colesbury**
t.as_param() is a drop-in replacement for the call Variable(t.data, requires_grad=True).

### CODE REVIEW

**Comment by vadimkantorov** [comment]
should this be named `attach()` (if `detach` still exists...)? https://github.com/pytorch/pytorch/issues/1734

**Comment by ezyang** [comment]
@vadimkantorov We decided that we all really don't like the name `detach`. No one knows what it does.

**Review by gchanan — APPROVED** [review]

**Comment by apaszke** [comment]
Can we please find a different name? When I saw `as_param()` I initially though it returns an `nn.Parameter`

File: torch/csrc/autograd/variable.cpp [hunk 1]
```diff
@@ -164,6 +164,12 @@ void Variable::detach_() {
   set_gradient_edge(Edge());
 }
 
+Variable Variable::as_param() const {
+  auto res = make_variable(data(), /*requires_grad=*/true);
+  res.set_version_counter(version_counter());
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: I don't think what we're doing here is safe. If you do `v.as_param().mul_(2)`, then this won't get registered in the original graph, and the derivative will be incorrect!

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/csrc/autograd/variable.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: Couldn't we make it a view that requires grad?

Comment (SsnL) [reply torch/csrc/autograd/variable.cpp]: Why wouldn't this be tracked? This is the exact `y = x.detach(); y.required_grad = True` syntax. It's sharing the same version counter with the original tensor, so any inplace operation that breaks backward should be detected.

Comment (SsnL) [reply torch/csrc/autograd/variable.cpp]: It is also safer than `Variable(x.data, requires_grad=True)`.

Comment (SsnL) [reply torch/csrc/autograd/variable.cpp]: Oh I misread your comment. I'm sorry. Yeah, the concern is valid.

**Comment by soumith** [comment]
after offline discussion, we're going with `start_grad` (attach) / `stop_grad` (detach)

**Comment by themightyoarfish** [comment]
What is the status of this? Doesn't seem like anything made it into `master`.

**Comment by SsnL** [comment]
@themightyoarfish It was not. We decided to go with `requires_grad_` instead.
